test/csr/csr.cc

```diff
@@ -0,0 +1,259 @@
+/*
+ * csr.cc
+ *
+ * Copyright (C) 2017-2024 Tactical Computing Laboratories, LLC
+ * All Rights Reserved
+ * contact@tactcomplabs.com
+ *
+ * See LICENSE in the top level directory for licensing details
+ *
+ */
+
+#include "syscalls.h"
+#include <inttypes.h>
+#include <stddef.h>
+#include <stdio.h>
+#include <string.h>
+
+#define assert( x )      \
+  do {                   \
+    if( !x )             \
+      asm( " .word 0" ); \
+  } while( 0 )
+
+enum class CSROp { Write, Set, Clear };
+
+enum class OpKind { Imm, Reg };
+
+template<CSROp OP, OpKind KIND, uint16_t CSR, size_t ARG, bool RD0>
+struct CSRInst;
+
+template<uint16_t CSR, size_t ARG, bool RD0>
+struct CSRInst<CSROp::Write, OpKind::Reg, CSR, ARG, RD0> {
+  size_t operator()() {
+    size_t old = 0;
+    if constexpr( RD0 ) {
+      if constexpr( ARG == 0 ) {
+        asm volatile( "csrrw zero, %0, zero" ::"I"( CSR ) );
+      } else {
+        size_t arg = ARG;
+        asm volatile( "csrrw zero, %0, %1" ::"I"( CSR ), "r"( arg ) );
+      }
+    } else {
+      if constexpr( ARG == 0 ) {
+        asm volatile( "csrrw %0, %1, zero" : "=r"( old ) : "I"( CSR ) );
+      } else {
+        size_t arg = ARG;
+        asm volatile( "csrrw %0, %1, %2" : "=r"( old ) : "I"( CSR ), "r"( arg ) );
+      }
+    }
+    return old;
+  }
+};
+
+template<uint16_t CSR, size_t ARG, bool RD0>
+struct CSRInst<CSROp::Set, OpKind::Reg, CSR, ARG, RD0> {
+  size_t operator()() {
+    size_t old = 0;
+    if constexpr( RD0 ) {
+      if constexpr( ARG == 0 ) {
+        asm volatile( "csrrs zero, %0, zero" ::"I"( CSR ) );
+      } else {
+        size_t arg = ARG;
+        asm volatile( "csrrs zero, %0, %1" ::"I"( CSR ), "r"( arg ) );
+      }
+      return old;
+    } else {
+      if constexpr( ARG == 0 ) {
+        asm volatile( "csrrs %0, %1, zero" : "=r"( old ) : "I"( CSR ) );
+      } else {
+        size_t arg = ARG;
+        asm volatile( "csrrs %0, %1, %2" : "=r"( old ) : "I"( CSR ), "r"( arg ) );
+      }
+      return old;
+    }
+  }
+};
+
+template<uint16_t CSR, size_t ARG, bool RD0>
+struct CSRInst<CSROp::Clear, OpKind::Reg, CSR, ARG, RD0> {
+  size_t operator()() {
+    size_t old = 0;
+    if constexpr( RD0 ) {
+      if constexpr( ARG == 0 ) {
+        asm volatile( "csrrc zero, %0, zero" ::"I"( CSR ) );
+      } else {
+        size_t arg = ARG;
+        asm volatile( "csrrc zero, %0, %1" ::"I"( CSR ), "r"( arg ) );
+      }
+    } else {
+      if constexpr( ARG == 0 ) {
+        asm volatile( "csrrc %0, %1, zero" : "=r"( old ) : "I"( CSR ) );
+      } else {
+        size_t arg = ARG;
+        asm volatile( "csrrc %0, %1, %2" : "=r"( old ) : "I"( CSR ), "r"( arg ) );
+      }
+    }
+    return old;
+  }
+};
```

Comment:
When the operand to the CSR instruction is a register, we store the value in a register and pass that register to the CSR instruction. If the value is `0`, we use the `x0` register instead of another register which happens to be `0`. When `rd == x0`, we explicitly code `rd` as `x0`.